WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemUiStateHelper.kt

```diff
@@ -117,6 +132,38 @@ class PageItemUploadProgressHelper @Inject constructor(
         }
     }
 
+    fun setupPageActions(listType: PageListType, pageId: LocalId, site: SiteModel): Set<Action> {
+        return when (listType) {
+            SCHEDULED, PUBLISHED -> mutableSetOf(VIEW_PAGE, SET_PARENT, MOVE_TO_DRAFT, MOVE_TO_TRASH).apply {
+                if (canCancelPendingAutoUpload(pageId, site)) {
+                    add(CANCEL_AUTO_UPLOAD)
+                }
+            }
+            DRAFTS -> mutableSetOf(VIEW_PAGE, SET_PARENT, PUBLISH_NOW, MOVE_TO_TRASH).apply {
+                if (canCancelPendingAutoUpload(pageId, site)) {
+                    add(CANCEL_AUTO_UPLOAD)
+                }
+            }
+            TRASHED -> setOf(MOVE_TO_DRAFT, DELETE_PERMANENTLY)
+        }
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    fun canCancelPendingAutoUpload(pageId: LocalId, site: SiteModel): Boolean {
+        val post = postStore.getPostByLocalPostId(pageId.value)
```

Comment:
I'd suggest adding a TODO here, since this is yet another place which we'll need to update as the current implementation would have significant performance impact.

Reply:
Resolved here e036fa0